src/libopensc/jacartapki.c

```diff
@@ -0,0 +1,1643 @@
+	return;
+}
+
+static int
+_cka_get_unsigned(const struct jacartapki_cka *attr, unsigned *out)
+{
+	int ii;
+
+	if (attr == NULL || out == NULL)
+		return SC_ERROR_INVALID_ARGUMENTS;
+	if (attr->len != 4)
+		return SC_ERROR_INVALID_DATA;
+
+	for (ii = 0, *out = 0; ii < 4; ii++)
+		*out = (*out << 8) + attr->val[3 - ii];
+
+	return SC_SUCCESS;
+}
+
+static int
+_cka_set_label(const struct jacartapki_cka *attr, struct sc_pkcs15_object *obj)
+{
+	size_t len;
+
+	if (attr == NULL || obj == NULL)
+		return SC_ERROR_INVALID_ARGUMENTS;
+
+	memset(obj->label, 0, sizeof(obj->label));
+	len = MIN(attr->len, sizeof(obj->label) - 1);
+	if (len > 0)
+		memcpy(obj->label, attr->val, len);
+
+	return SC_SUCCESS;
+}
+
+static int
+_cka_set_application(const struct jacartapki_cka *attr, struct sc_pkcs15_data_info *info)
+{
+	size_t len;
+
+	if (!attr || !info)
+		return SC_ERROR_INVALID_ARGUMENTS;
+
+	memset(info->app_label, 0, sizeof(info->app_label));
+	len = MIN(attr->len, sizeof(info->app_label) - 1);
+	if (len)
+		memcpy(info->app_label, attr->val, len);
+
+	return SC_SUCCESS;
+}
+
+static int
+_cka_get_object_id(const struct jacartapki_cka *attr, struct sc_pkcs15_data_info *info)
+{
+	int ii;
+
+	for (ii = 0; ii < SC_MAX_OBJECT_ID_OCTETS; ii++) {
+		if (ii * sizeof(int) < attr->len)
+			info->app_oid.value[ii] = *((int *)(attr->val + ii * sizeof(int)));
+		else
+			info->app_oid.value[ii] = -1;
+	}
+
+	return SC_SUCCESS;
+}
+
+static int
+_cka_get_blob(const struct jacartapki_cka *attr, struct sc_pkcs15_der *out)
+{
+	struct sc_pkcs15_der der;
+
+	if (attr == NULL || out == NULL)
+		return SC_ERROR_INVALID_ARGUMENTS;
+
+	der.value = malloc(attr->len);
+	if (der.value == NULL)
+		return SC_ERROR_OUT_OF_MEMORY;
+	memcpy(der.value, attr->val, attr->len);
+	der.len = attr->len;
+
+	*out = der;
+	return SC_SUCCESS;
+}
+
+static int
+_cka_set_id(const struct jacartapki_cka *attr, struct sc_pkcs15_id *out)
+{
+	if (attr == NULL || out == NULL)
+		return SC_ERROR_INVALID_ARGUMENTS;
+
+	if (attr->len > SC_PKCS15_MAX_ID_SIZE)
+		return SC_ERROR_INVALID_DATA;
+
+	memcpy(out->value, attr->val, attr->len);
+	out->len = attr->len;
+
+	return SC_SUCCESS;
+}
+
+static int
+jacartapki_add_attribute(unsigned char **buf, size_t *buf_sz, unsigned char flags,
+		CK_ULONG cka, size_t cka_len, const void *data)
+{
+	unsigned char *ptr = NULL;
+	size_t offs = 0;
+
+	if (buf == NULL || buf_sz == NULL)
+		return SC_ERROR_INVALID_ARGUMENTS;
+
+	ptr = realloc(*buf, *buf_sz + cka_len + 5);
+	if (ptr == NULL)
+		return SC_ERROR_OUT_OF_MEMORY;
+
+	offs = *buf_sz;
+	if (cka & CKA_VENDOR_DEFINED)
+		ptr[offs++] = (cka >> 24) & 0xFF; /* cka type: MSB | LSB */
+	else
+		ptr[offs++] = (cka >> 8) & 0xFF; /* cka type: 2 LSBs */
+	ptr[offs++] = cka & 0xFF;
+	ptr[offs++] = flags;
+	ptr[offs++] = (cka_len >> 8) & 0xFF; /* cka length: 2 bytes*/
+	ptr[offs++] = cka_len & 0xFF;
+
+	memset(ptr + offs, 0, cka_len);
+	if (data != NULL)
+		memcpy(ptr + offs, (const unsigned char *)data, cka_len);
```

Comment:
afaik this should be enough to memset only if we do not do memcpy:
```suggestion
	if (data != NULL)
		memcpy(ptr + offs, (const unsigned char *)data, cka_len);
	else
		memset(ptr + offs, 0, cka_len);
```